### PR TITLE
refactor: move CTAS/CVAS field

### PR DESCRIPTION
**Pull request by lyndsiWilliams**
### SUMMARY
Moved CTAS/CVAS Schema Field Closer to the "Allow Create Table As" In Edit Datasource Modal. This required nested conditional rendering, so I had to use some useState/useEffect in order to hold that state at a higher level for all functions to run properly.

#### Field before:
ctascvas-before

#### Field after:
ctascvas-after-1
ctascvas-after-2
ctascvas-after-3

### TEST PLAN
Testing has not yet been implemented for this.

### ADDITIONAL INFORMATION

- [ ] Has associated issue:
- [x] Changes UI
- [ ] Requires DB Migration.
- [ ] Confirm DB Migration upgrade and downgrade tested.
- [ ] Introduces new feature or API
- [ ] Removes existing feature or API

### CODE REVIEW

File: superset-frontend/src/views/CRUD/data/database/DatabaseModal.tsx [hunk 1]
```diff
@@ -235,6 +249,36 @@ const DatabaseModal: FunctionComponent<DatabaseModalProps> = ({
...
+      setCount(count + 1);
+    } else {
+      setCount(count - 1);
+    }
+  };
+
+  useEffect(() => {
+    // Opens the form if at least one option is checked
+    // Closes the form only if both options are unchecked
+    if (count > 0) {
```
Comment (hughhhh) [inline superset-frontend/src/views/CRUD/data/database/DatabaseModal.tsx:266]: couldn't this just a be a boolean if `expandForm` is called for `Expose SQL lab`

File: superset-frontend/src/views/CRUD/data/database/DatabaseModal.tsx [hunk 2]
```diff
@@ -467,87 +535,101 @@ const DatabaseModal: FunctionComponent<DatabaseModalProps> = ({
                   tooltip={t('Allow this database to be queried in SQL Lab')}
                 />
               </div>
+              <StyledExpandableForm
```
Comment (hughhhh) [inline superset-frontend/src/views/CRUD/data/database/DatabaseModal.tsx:538]: Can we add some test for this component to make sure it interacts properly

File: superset-frontend/src/views/CRUD/data/database/DatabaseModal.tsx [hunk 3]
```diff
@@ -235,6 +249,36 @@ const DatabaseModal: FunctionComponent<DatabaseModalProps> = ({
     }
   };
 
+  const expandForm = (trueOrFalse: boolean) => {
```
Comment (eschutho) [inline superset-frontend/src/views/CRUD/data/database/DatabaseModal.tsx:252]: small nit, can we name this arg something like `shouldExpand`?

File: superset-frontend/src/views/CRUD/data/database/DatabaseModal.tsx [hunk 4]
```diff
@@ -235,6 +249,36 @@ const DatabaseModal: FunctionComponent<DatabaseModalProps> = ({
...
+    // Count starts a 0 (closed)
+    // Any checkbox in this for that gets checked
+    // will increase the count and keep the form open
+    if (trueOrFalse) {
+      setCount(count + 1);
+    } else {
+      setCount(count - 1);
+    }
+  };
```
Comment (eschutho) [inline superset-frontend/src/views/CRUD/data/database/DatabaseModal.tsx:261]: maybe more succinct:
`setCount(count + (shouldExpand ? 1 : -1)`

File: superset-frontend/src/views/CRUD/data/database/DatabaseModal.tsx [hunk 5]
```diff
@@ -235,6 +249,36 @@ const DatabaseModal: FunctionComponent<DatabaseModalProps> = ({
...
+
+  useEffect(() => {
+    // Opens the form if at least one option is checked
+    // Closes the form only if both options are unchecked
+    if (count > 0) {
+      setCreateAsOpen(true);
+    } else {
+      setCreateAsOpen(false);
+    }
```
Comment (eschutho) [inline superset-frontend/src/views/CRUD/data/database/DatabaseModal.tsx:270]: also more succinct: 
`setCreateAsOpen(count>0)`

File: superset-frontend/src/views/CRUD/data/database/DatabaseModal.tsx [hunk 6]
```diff
@@ -250,6 +294,30 @@ const DatabaseModal: FunctionComponent<DatabaseModalProps> = ({
         typeof target.value === 'string' ? target.value.trim() : target.value;
     }
 
+    if (target.id === 'expose_in_sqllab') {
+      setIsOpen(!isOpen);
+    }
+
+    // Conditional form rendering
+    const checkedTrue = target.checked === true;
```
Comment (eschutho) [inline superset-frontend/src/views/CRUD/data/database/DatabaseModal.tsx:302]: `const checkedTrue = target.checked` should work.

Comment (eschutho) [reply superset-frontend/src/views/CRUD/data/database/DatabaseModal.tsx:302]: actually even better = `const { checked } = target`

File: superset-frontend/src/views/CRUD/data/database/DatabaseModal.tsx [hunk 7]
```diff
@@ -250,6 +294,30 @@ const DatabaseModal: FunctionComponent<DatabaseModalProps> = ({
...
+    /* The CTAS & CVAS schema field needs individual setters
+    so that the input will not disappear when one options goes
+    unchecked but the other is still checked */
+    if (target.id.includes('cvas')) {
+      if (!cvas && checkedTrue) {
+        cvasCtasCheck('cvas', true);
+      } else {
+        cvasCtasCheck('cvas', false);
+      }
```
Comment (eschutho) [inline superset-frontend/src/views/CRUD/data/database/DatabaseModal.tsx:311]: shorter: 
```
const { id, checked } = target;
const check = (!cvas && id.includes('cvas') && 'cvas') || (!ctas && id.includes('ctas') && 'ctas');
cvasCtasCheck( check, checked);
```

File: superset-frontend/src/views/CRUD/data/database/DatabaseModal.tsx [hunk 8]
```diff
@@ -467,87 +535,101 @@ const DatabaseModal: FunctionComponent<DatabaseModalProps> = ({
...
               </div>
+              <StyledExpandableForm
+                style={{ display: isOpen ? 'inherit' : 'none' }}
+              >
+                <StyledInputContainer>
+                  <div className="input-container">
+                    <IndeterminateCheckbox
+                      id="allow_ctas"
+                      indeterminate={false}
+                      checked={db ? !!db.allow_ctas : false}
```
Comment (eschutho) [inline superset-frontend/src/views/CRUD/data/database/DatabaseModal.tsx:546]: `checked={db && !!db.allow_ctas}`

File: superset-frontend/src/views/CRUD/data/database/DatabaseModal.tsx [hunk 9]
```diff
@@ -467,87 +535,101 @@ const DatabaseModal: FunctionComponent<DatabaseModalProps> = ({
...
+                      )}
+                    />
+                  </div>
+                </StyledInputContainer>
+                <StyledInputContainer>
+                  <div className="input-container">
+                    <IndeterminateCheckbox
+                      id="allow_cvas"
+                      indeterminate={false}
+                      checked={db ? !!db.allow_cvas : false}
```
Comment (eschutho) [inline superset-frontend/src/views/CRUD/data/database/DatabaseModal.tsx:562]: similar to line 546

File: superset-frontend/src/views/CRUD/data/database/DatabaseModal.tsx [hunk 10]
```diff
@@ -467,87 +535,101 @@ const DatabaseModal: FunctionComponent<DatabaseModalProps> = ({
...
+                    style={{ display: createAsOpen ? 'inherit' : 'none' }}
+                  >
+                    <div className="control-label">
+                      {t('CTAS & CVAS SCHEMA')}
+                    </div>
+                    <div className="input-container">
+                      <input
+                        type="text"
+                        name="force_ctas_schema"
+                        value={db ? db.force_ctas_schema || '' : ''}
```
Comment (eschutho) [inline superset-frontend/src/views/CRUD/data/database/DatabaseModal.tsx:582]: same refactor as above

**Comment by eschutho** [comment]
Looks great. I just did a syntax check. I can look again for overall logic/layout.

File: superset-frontend/src/views/CRUD/data/database/DatabaseModal.tsx [hunk 11]
```diff
@@ -120,6 +120,15 @@ const StyledJsonEditor = styled(JsonEditor)`
   border-radius: ${({ theme }) => theme.gridUnit}px;
 `;
 
+const StyledExpandableForm = styled.div`
+  padding-left: 28px;
```
Comment (mistercrunch) [inline superset-frontend/src/views/CRUD/data/database/DatabaseModal.tsx:124]: let's use `theme.gridUnit` as above with the right multiplier

File: superset-frontend/src/views/CRUD/data/database/DatabaseModal.tsx [hunk 12]
```diff
@@ -132,6 +141,11 @@ const DatabaseModal: FunctionComponent<DatabaseModalProps> = ({
   const [db, setDB] = useState<DatabaseObject | null>(null);
   const [isHidden, setIsHidden] = useState<boolean>(true);
   const [tabKey, setTabKey] = useState<string>(DEFAULT_TAB_KEY);
+  const [isOpen, setIsOpen] = useState<boolean>(true);
```
Comment (nytai) [inline superset-frontend/src/views/CRUD/data/database/DatabaseModal.tsx:144]: does this refer to the modal being open or something else?

**Comment by eschutho** [comment]
@lyndsiWilliams do you want to put this in draft mode for the time being?

**Pull request closed without merging**